src/core/skills/characters/yijiang2013/longyin.ts

```diff
@@ -0,0 +1,117 @@
+import { Sanguosha } from 'core/game/engine';
+import { GameCommonRules } from 'core/game/game_rules';
+import { AllStage, CardUseStage, PhaseChangeStage, PlayerPhase } from 'core/game/stage_processor';
+import { Player } from 'core/player/player';
+import { PlayerId } from 'core/player/player_props';
+import { Room } from 'core/room/room';
+import { CommonSkill, TriggerSkill } from 'core/skills/skill';
+
+@CommonSkill({ name: 'longyin', description: 'longyin_description' })
+export class LongYin extends TriggerSkill {
+  public isTriggerable(
+    event: ServerEventFinder<GameEventIdentifiers.CardUseEvent | GameEventIdentifiers.PhaseChangeEvent>,
+    stage?: AllStage,
+  ) {
+    return stage === CardUseStage.CardUsing || stage === PhaseChangeStage.AfterPhaseChanged;
+  }
+
+  isAutoTrigger(
+    room: Room,
+    owner: Player,
+    event: ServerEventFinder<GameEventIdentifiers.CardUseEvent | GameEventIdentifiers.PhaseChangeEvent>,
+  ) {
+    const identifier = EventPacker.getIdentifier(event);
+    return identifier === GameEventIdentifiers.PhaseChangeEvent;
+  }
+
+  public canUse(
+    room: Room,
+    owner: Player,
+    content: ServerEventFinder<GameEventIdentifiers.CardUseEvent | GameEventIdentifiers.PhaseChangeEvent>,
+  ) {
+    const identifier = EventPacker.getIdentifier(content);
+    if (identifier === GameEventIdentifiers.CardUseEvent) {
+      const event = content as ServerEventFinder<GameEventIdentifiers.CardUseEvent>;
+      return (
+        room.CurrentPhasePlayer.Id === event.fromId &&
+        Sanguosha.getCardById(event.cardId).GeneralName === 'slash' &&
+        owner.getPlayerCards().length > 0
+      );
+    } else {
+      const event = content as ServerEventFinder<GameEventIdentifiers.PhaseChangeEvent>;
+      return event.to === PlayerPhase.PhaseFinish && owner.getFlag<PlayerId>(this.Name) !== undefined;
+    }
+  }
+
+  cardFilter(room: Room, owner: Player, cards: CardId[]) {
+    return cards.length === 1;
+  }
+
+  public isAvailableCard(): boolean {
+    return true;
+  }
+
+  public async onTrigger(room: Room, event: ServerEventFinder<GameEventIdentifiers.SkillUseEvent>) {
+    const { triggeredOnEvent } = event;
+    const identifier = EventPacker.getIdentifier(triggeredOnEvent!);
+    if (identifier === GameEventIdentifiers.PhaseChangeEvent) {
+      event.translationsMessage = undefined;
+    }
+    return true;
+  }
+
+  public async onEffect(room: Room, skillUseEvent: ServerEventFinder<GameEventIdentifiers.SkillEffectEvent>) {
+    const { fromId, cardIds, triggeredOnEvent } = skillUseEvent;
+    const from = room.getPlayerById(fromId);
+    const identifier = EventPacker.getIdentifier(triggeredOnEvent!);
+    if (identifier === GameEventIdentifiers.PhaseChangeEvent) {
+      const targetId = from.getFlag<PlayerId>(this.Name);
+      const extraUse = from.getInvisibleMark(this.Name);
+      room.syncGameCommonRules(targetId, target => {
```

Comment:
As my understand, you will set the mark when you invoke LongYin, and clear the mark at FinishPhase;
when invoke LongYin, you add the slash usable Time by 1, so player can use slash time is the sum of (Base + MarkNum), and you clear the mark at FinishPhase;

There are some traps:
1. If Presenter is PlayerA, she give other PlayerB a PlayPhase, when you invoke this skill on PlayerB's Phase, you can't clear you mark Because PlayerB Don't Have a FinishPhase,  so PlayerB can use (BASE+MarkNum) slash at least;
2. If Presenter has multipe PlayPhase such as LiaoHua, she can use (Base+MarkNum) slash at least in his other PlayPhase;
And I Don't know what will happen if LongYin's Owner died but not clear the mark;

Reply:
yea, that's my bad, should be cleared after each play card stage ended